crates/libs/windows/src/core/heap.rs

```diff
@@ -32,20 +32,42 @@ pub unsafe fn heap_free(ptr: RawPtr) {
     }
 }
 
-/// Copy a slice of `T` into a freshly allocated buffer with an additional default `T` at the end.
+/// Copy an iterator of `T` into a freshly allocated buffer with an additional default `T` at the end.
 ///
-/// Returns a pointer to the beginning of the buffer
+/// Returns a pointer to the beginning of the buffer. This pointer must be freed when done using `heap_free`.
 ///
 /// # Panics
 ///
-/// This function panics if the heap allocation fails or if the pointer returned from
-/// the heap allocation is not properly aligned to `T`.
-pub fn heap_string<T: Copy + Default + Sized>(slice: &[T]) -> *const T {
-    unsafe {
-        let buffer = heap_alloc((slice.len() + 1) * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
-        assert!(buffer.align_offset(std::mem::align_of::<T>()) == 0, "heap allocated buffer is not properly aligned");
-        buffer.copy_from_nonoverlapping(slice.as_ptr(), slice.len());
-        buffer.add(slice.len()).write(T::default());
-        buffer
+/// This function panics if the heap allocation fails, the alignment requirements of 'T' surpass
+/// 8 (HeapAlloc's alignment) or if len is less than the number of items in the iterator.
+pub fn string_from_iter<I, T>(iter: I, len: usize) -> *const T
+where
+    I: Iterator<Item = T>,
+    T: Copy + Default,
+{
+    // alignment of memory returned by HeapAlloc is at least 8
+    // Source: https://docs.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc
+    // Ensure that T has sufficient alignment requirements
+    assert!(std::mem::align_of::<T>() <= 8, "T alignment surpasses HeapAlloc alignment");
+
+    let len = len + 1;
+    let ptr = heap_alloc(len * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
+    let mut encoder = iter.chain(core::iter::once(T::default()));
+
+    for i in 0..len {
+        // SAFETY: ptr points to an allocation object of size `len`, indices accessed are always lower than `len`
+        unsafe {
+            core::ptr::write(
+                ptr.add(i),
+                match encoder.next() {
+                    Some(encoded) => encoded,
+                    None => break,
+                },
+            );
+        }
```

Comment:
Make this a bit clearer perhaps:

```rust
for (offset, c) in (0..len).zip(encoder) {
  // SAFETY: ptr points to an allocation object of size `len`, indices accessed are always lower than `len`
  unsafe { core::ptr::write(ptr.add(offset), c); }
}
```

Reply:
This looks much better! Unfortunately putting the encoder into `.zip(self)` would consume it, which wouldn't allow our assertion afterwards:

```rust
assert!(encoder.next().is_none(), "encoder returned more characters than expected");
```

Your version does look much better and I'd take it in a heartbeat, but we'd need a non-consuming version of zip unfortunately.

Reply:
It would be unusual, but is it invalid to request encoding fewer characters than there are?

Reply:
Technically in my opinion it is impossible, but since there was uncertainty in the reviews about silent truncation I kept it in. It could be possible that somehow an invalid length gets passed in the future, so it'd be helpful to be aware via a panic rather than silent truncation. I'm not hugely opinionated on this issue, we can go for the more elegant `.zip` at the expense of potential bugs in the encoder or what gets passed to the function.

Reply:
I wasn't worried about truncation, I was worried about writing past the allocated memory with the unsafe `ptr.write()`.  Your outer loop and/or the `.zip()` makes that a non-issue.  No strong opinion on this either, I can see it both ways.  We're not going to overwrite anything no matter what.

Reply:
The `Zip` iterator instance will return `None` because the range iterator is finished.  You can never get to the original `encoder` iterator again because it's inside the `Zip` iterator instance.